syntax = "proto3";

package chat;

// Account management messages
message UsernameRequest {
  string username = 1;
}

message UsernameResponse {
  bool exists = 1;
  string message = 2;
}

message CreateAccountRequest {
  string username = 1;
  string hashed_password = 2;
}

message LoginRequest {
  string username = 1;
  string hashed_password = 2;
}

message AuthResponse {
  bool success = 1;
  string username = 2;
  string message = 3;
  int32 unread_count = 4;
}

message LogoutRequest {
  string username = 1;
}

message StatusResponse {
  bool success = 1;
  string message = 2;
}

// Messaging messages
message ListAccountsRequest {
  string username = 1;
  string pattern = 2;
}

message ListAccountsResponse {
  bool success = 1;
  repeated string accounts = 2;
  string message = 3;
}

message SendMessageRequest {
  string sender = 1;
  string recipient = 2;
  string content = 3;
}

message ViewMessagesRequest {
  string username = 1;
  int32 count = 2;
}

message ViewMessagesResponse {
  bool success = 1;
  repeated MessageData messages = 2;
  string message = 3;
}

message MessageData {
  string sender = 1;
  string content = 2;
  string timestamp = 3;
}

message MessageResponse {
  string sender = 1;
  string content = 2;
  string timestamp = 3;
}

message DeleteMessagesRequest {
  string username = 1;
  string delete_info = 2; // "ALL" or a number as string
}

// Replication message
message ReplicationRequest {
  string operation = 1;     // e.g., "CreateAccount", "SendMessage", "DeleteMessages"
  string data = 2;          // JSON-encoded update details
  string update_id = 3;     // Unique update identifier
}

// Leader election / heartbeat messages
message PingRequest {
  string dummy = 1;
}

message PingResponse {
  string message = 1;
  string sender_id = 2;
}

message LeaderRequest {
  // Empty
}

message LeaderResponse {
  string leader_address = 1;
}

// Service definition
service ChatService {
  // Account management
  rpc CheckUsername(UsernameRequest) returns (UsernameResponse);
  rpc CreateAccount(CreateAccountRequest) returns (AuthResponse);
  rpc Login(LoginRequest) returns (AuthResponse);
  rpc Logout(LogoutRequest) returns (StatusResponse);
  rpc DeleteAccount(UsernameRequest) returns (StatusResponse);
  
  // Messaging operations
  rpc ListAccounts(ListAccountsRequest) returns (ListAccountsResponse);
  rpc SendMessage(SendMessageRequest) returns (StatusResponse);
  rpc ViewMessages(ViewMessagesRequest) returns (ViewMessagesResponse);
  rpc DeleteMessages(DeleteMessagesRequest) returns (StatusResponse);
  
  // Streaming for real-time message delivery
  rpc ReceiveMessages(UsernameRequest) returns (stream MessageResponse);
  
  // Replication RPC
  rpc ReplicateOperation(ReplicationRequest) returns (StatusResponse);
  
  // Leader election and heartbeat RPCs
  rpc Ping(PingRequest) returns (PingResponse);
  rpc GetLeader(LeaderRequest) returns (LeaderResponse);
}